src/stim/stabilizers/pauli_string_iter.inl

```diff
@@ -0,0 +1,306 @@
+    t1 = cur_perm;
+    size_t num_zeros = count_trailing_zeros(cur_perm);
+    for (size_t w = 0; w < cur_perm.num_u64_padded(); w++) {
+        // Either we have an all all ones or an all zeros integer in which case we
+        // skip, otherwise this word must contain the least significant bit
+        if (cur_perm.u64[w] != ones_bit && cur_perm.u64[w]) {
+            // Use the single word algorithm from above
+            t1.u64[w] = cur_perm.u64[w] | ((1ULL << num_zeros) - 1);
+            // Set all previous words to ones
+            for (size_t old_w = 0; old_w < w; old_w++) {
+                t1.u64[old_w] = ones_bit;
+            }
+            break;
+        }
+    }
+    // invert the bits
+    for (size_t w = 0; w < cur_perm.num_u64_padded(); w++) {
+        if (t1.u64[w]) {
+            t1_inv.u64[w] = ~t1.u64[w];
+        } else {
+            // all zeros so set to all ones
+            t1_inv.u64[w] = ones_bit;
+        }
+    }
+    size_t next_zero = count_trailing_zeros(t1_inv);
+    for (size_t w = 0; w < cur_perm.num_u64_padded(); w++) {
+        if (next_zero / 64 == w) {
+            if ((next_zero + 1) % 64 == 0) {
+                t2.u64[w] = ones_bit;
+            } else {
+                t2.u64[w] = ((1ULL << (next_zero + 1 - 64 * w)) - 1);
+            }
+            break;
+        } else {
+            t2.u64[w] = ones_bit;
+        }
+    }
+    size_t word_loc = (next_zero - num_zeros - 1) / 64;
+    for (size_t w = 0; w < word_loc; w++) {
+        t3.u64[w] = ones_bit;
+    }
+    // next_zero - num_zeros - 1 - words * 64 is always < 64, so don't need to worry about overflow like above.
+    t3.u64[word_loc] = (1ULL << (next_zero - num_zeros - 1 - word_loc * 64)) - 1;
+    cur_perm = t1;
+    cur_perm ^= t2;
+    cur_perm |= t3;
+}
+
+template <size_t W>
+bool PauliStringIterator<W>::iter_all_cur_perm(std::vector<int> &set_bits) {
+    size_t num_terms = static_cast<size_t>(pow(3, cur_w));
+    result.xs.clear();
+    result.zs.clear();
+    // Set to all Xs initially
+    for (size_t i = 0; i < cur_w; i++) {
+        result.xs.u64[set_bits[i] / 64] ^= uint64_t{1} << (set_bits[i] & 63);
+    }
+    // TODO: in principle we could store the 3^w patterns for reasonable values
+    // of w and avoid repeating this loop for each new qubit permutation. We
+    // would save a mod and division so probably not worth it.
+    while (cur_k < num_terms) {
+        int n = cur_k;
+        int indx = 0;
+        while (n > 0) {
+            // Use ternary representation of cur_k to provide XZY (0, 1, 2) Pauli pattern
+            // Recall the ternary ordering for cur_w = 2, num_qubits = 2 would yield
+            // 00 = XX
+            // 01 = XZ
+            // 02 = XY
+            // 10 = YX
+            // 11 = YZ
+            // 12 = YY  ...
+            // This inner while loop will yield r and indx, where r is the
+            // trit (0, 1, 2) and indx is the corresponding qubit index.
+            int r = n % 3;
```

Comment:
If you use adjacent bits in a word to encode a pauli (00=I, 01=X, ...) then you can use `a++` followed by `a |= ~(a | (a >> 1)) & 0x5555555555555555ULL` to skip over values that would have identity paulis.

```
def bits_to_paulis(x: int, n: int) -> str:
    bits = bin(x)[2:].rjust(n*2, '0')[-n*2:]
    return ''.join('_XYZ'[int(bits[k:k+2], 2)] for k in range(0, len(bits), 2))


a = 0
for _ in range(3**3):
    a += 1
    a |= ~(a | (a >> 1)) & 0x5555555555555555
    print(bits_to_paulis(a, 4))
```

prints:

```
XXXX
XXXY
XXXZ
XXYX
XXYY
XXYZ
XXZX
XXZY
XXZZ
XYXX
XYXY
XYXZ
XYYX
XYYY
XYYZ
XYZX
XYZY
XYZZ
XZXX
XZXY
XZXZ
XZYX
XZYY
XZYZ
XZZX
XZZY
XZZZ
```

Reply:
resolved with pair_sat_increment.